core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java

```diff
@@ -170,6 +170,13 @@ private DataBuffer readChunkInternal() throws IOException {
       return null;
     }
     mPosToRead += buffer.readableBytes();
+    if (mPosToRead - mReadRequest.getOffset() == mReadRequest.getLength()) {
+      // finish reading all the data, close gRPC stream earlier to release resources
+      // potentially improve machine learning millions of small file read throughput
+      LOG.debug("Finished reading all data, close gRPC stream");
+      close();
+      return buffer;
+    }
     try {
       mStream.send(mReadRequest.toBuilder().setOffsetReceived(mPosToRead).build());
```

Comment:
unrelated to this PR, but could we send this receipt asynchronously to return the data to the caller faster? not sure if this would even make a difference, but just a thought.

Reply:
For the small file use case, they only read one chunk and no need to send offset received. Without the fixes in this PR, client will send the offset received to worker, but worker won't stop setting data until a gap larger than 4MB. Not sure is it possible or helpful to do it asynchronously, maybe @ggezer can have a better idea.

Reply:
This call would effectively be asynchronous by setting `alluxio.user.streaming.reader.close.timeout` to `0`.

Reply:
not familiar with this part of the code, but it seems to me that it is only asynchronous if there is one chunk to read right? if there are multiple chunks, is this send not in the critical path of returning the data to the caller?

Reply:
By asynchronous i meant, caller wouldn't have to wait at all since onCompleted() returns immediately. 
I agree that async execution of receipts would be another level of optimization.